libs/utils/src/leaky_bucket.rs

```diff
@@ -0,0 +1,225 @@
+        }
+        dur
+    }
+
+    fn next_multiple_of_drain(&self, mut dur: Duration) -> Duration {
+        if self.drain_interval > Duration::ZERO {
+            let n = dur.div_duration_f64(self.drain_interval).ceil();
+            dur = self.drain_interval.mul_f64(n);
+        }
+        dur
+    }
+}
+
+pub struct LeakyBucketState {
+    /// Bucket is represented by `start..end` where `end = epoch + end` and `start = end - config.bucket_width`.
+    ///
+    /// At any given time, `end - now` represents the number of tokens in the bucket, multiplied by the "time_cost".
+    /// Adding `n` tokens to the bucket is done by moving `end` forward by `n * config.time_cost`.
+    /// If `now < start`, the bucket is considered filled and cannot accept any more tokens.
+    /// Draining the bucket will happen naturally as `now` moves forward.
+    ///
+    /// Let `n` be some "time cost" for the request,
+    /// If now is after end, the bucket is empty and the end is reset to now,
+    /// If now is within the `bucket window + n`, we are within time budget.
+    /// If now is before the `bucket window + n`, we have run out of budget.
+    ///
+    /// This is inspired by the generic cell rate algorithm (GCRA) and works
+    /// exactly the same as a leaky-bucket.
+    pub end: Duration,
+}
+
+impl LeakyBucketState {
+    pub fn new(now: Duration) -> Self {
+        Self { end: now }
+    }
+
+    pub fn bucket_is_empty(&self, config: &LeakyBucketConfig, now: Instant) -> bool {
+        // if self.end is after now, the bucket is not empty
+        self.end <= config.prev_multiple_of_drain(now - config.epoch)
+    }
+
+    /// Immediately adds tokens to the bucket, if there is space.
+    ///
+    /// In a scenario where you are waiting for available rate,
+    /// rather than just erroring immediately, `started` corresponds to when this waiting started.
+    ///
+    /// `n` is the number of tokens that will be filled in the bucket.
+    ///
+    /// # Errors
+    ///
+    /// If there is not enough space, no tokens are added. Instead, an error is returned with the time when
+    /// there will be space again.
+    pub fn add_tokens(
+        &mut self,
+        config: &LeakyBucketConfig,
+        started: Instant,
+        n: f64,
+    ) -> Result<(), Instant> {
+        let now = Instant::now();
+
+        // round down to the last time we would have drained the bucket.
+        let now = config.prev_multiple_of_drain(now - config.epoch);
+        let started = config.prev_multiple_of_drain(started - config.epoch);
+
+        // invariant: started <= now
+        debug_assert!(started <= now);
+
+        // If the bucket was empty when we started our search, bump the end up accordingly.
+        let mut end = self.end;
+        if end < started {
+            end = started;
+        }
+
+        let n = config.cost.mul_f64(n);
+        let end_plus_n = end + n;
+        let start_plus_n = end_plus_n.saturating_sub(config.bucket_width);
```

Comment:
This variable name is super confusing to me. It suggests this is `start + n` but it actually is more like "when would we have needed to start to stay exactly within the budget limit". No?

Reply:
```suggestion
        let wait_until = end_plus_n.saturating_sub(config.bucket_width);
```